File: src/com/thoughtworks/ppt/LambdaDemo.java
package com.thoughtworks.ppt;

public class LambdaDemo {
    public static void main(String[] args) {
        AStrategy aStrategy = new AStrategy();
        Worker worker = new Worker(aStrategy);
        worker.work();
    }
}

interface Strategy {
    void doSth();
}

class AStrategy implements Strategy {
    @Override
    public void doSth() {
        System.out.println("Do A Strategy");
    }
}

class BStrategy implements Strategy {
    @Override
    public void doSth() {
        System.out.println("Do B Strategy");
    }
}

class Worker {

    private Strategy strategy;

    public Worker(Strategy strategy) {
        this.strategy = strategy;
    }

    public void work() {
        System.out.println("START");
        if (strategy != null) {
            strategy.doSth();
        }
        System.out.println("END");
    }

    public static void main(String[] args) {
        // 第一种实现
        Worker worker = new Worker(new AStrategy());

        // 匿名类实现
        Worker worker1 = new Worker(new Strategy() {
            @Override
            public void doSth() {
                System.out.println("Do A Strategy");
            }
        } );

        // lambda表达式实现
        // 本质上lambda表达式也是匿名内部类
        Worker worker2 = new Worker( ()-> {
                System.out.println("Do A Strategy");
        } );
    }
}

File: src/com/thoughtworks/ppt/LambdaDemo.kt
package com.thoughtworks.ppt


class WorkerKotlin(private val strategy: () -> Unit) {
    fun work() {
        println("START")
        strategy.invoke()
        println("END")
    }
}

fun main() {

    val worker1 = WorkerKotlin {
        println("Do A Strategy")
    }

    val bStrategy = {
        println("Do B Strategy")
    }
    val worker2 = WorkerKotlin(bStrategy)
    worker1.work()
    worker2.work()
}
